pennylane/tape/interfaces/autograd.py

```diff
@@ -107,26 +97,74 @@ def _update_trainable_params(self):
         self.trainable_params = trainable_params
         self._all_parameter_values = params
 
-    def get_parameters(self, trainable_only=True):  # pylint: disable=missing-function-docstring
-        self._update_trainable_params()
-        params = self._all_parameter_values
+    def get_parameters(self, trainable_only=True, return_arraybox=False):
+        """Return the parameters incident on the tape operations.
+
+        The returned parameters are provided in order of appearance
+        on the tape. By default, the returned parameters are wrapped in
+        an ``autograd.builtins.list`` container.
+
+        Args:
+            trainable_only (bool): if True, returns only trainable parameters
+            return_arraybox (bool): if True, the returned parameters are not
+                wrapped in an ``autograd.builtins.list`` container
+        Returns:
+            autograd.builtins.list or list: the corresponding parameter values
+
+        **Example**
+
+        .. code-block:: python
 
-        if trainable_only:
-            params = [
-                p
-                for idx, p in enumerate(self._all_parameter_values)
-                if idx in self.trainable_params
-            ]
+            with JacobianTape() as tape:
+                qml.RX(0.432, wires=0)
+                qml.RY(0.543, wires=0)
+                qml.CNOT(wires=[0, 'a'])
+                qml.RX(0.133, wires='a')
+                expval(qml.PauliZ(wires=[0]))
 
-        return autograd.builtins.list(params)
+        By default, all parameters are trainable and will be returned:
+
+        >>> tape.get_parameters()
+        [0.432, 0.543, 0.133]
+
+        Setting the trainable parameter indices will result in only the specified
+        parameters being returned:
+
+        >>> tape.trainable_params = {1} # set the second parameter as free
+        >>> tape.get_parameters()
+        [0.543]
+
+        The ``trainable_only`` argument can be set to ``False`` to instead return
+        all parameters:
+
+        >>> tape.get_parameters(trainable_only=False)
+        [0.432, 0.543, 0.133]
+        """
+        params = []
+        iterator = self.trainable_params if trainable_only else self._par_info
+
+        for p_idx in iterator:
+            op = self._par_info[p_idx]["op"]
+            op_idx = self._par_info[p_idx]["p_idx"]
+            params.append(op.data[op_idx])
+
+        return params if return_arraybox else autograd.builtins.list(params)

```

Comment:
I think I'm not seeing the subtlety: what was the motivation for updating `get_parameters()` and `_update_trainable_params()`?

Reply:
Previously, `get_parameters()` would call `_update_trainable_params()` with every call. I don't think in retrospect this was the right approach;

* setting the trainable parameters should be called in general less often than `get_parameters()` -- the dependency should go the other way.

* It resulted in a `get_parameter()` method that differed wildly from the base tape method.

In this PR, this has been corrected:

* Now, `_update_trainable_params()` calls `get_parameters()`, to get the latest parameters and determine which are trainable.

* The contents of the `get_parameters()` method is now almost identical to the base method, with the exception of the final return statement.

The latter approach is the better approach with less overhead, with the subtlety that `_update_trainable_params()` needs to get the non-autograd-list sequence of parameters to determine trainability. So the latter approach should have been implemented from the start.

As to why it was changed in this PR; I can't answer that properly. The change below to `_execute` (to fix the bug) caused the old approach to stop working -- through trial and error I discovered that this approach (the better one) _does_ work with the change, and the entire PL test suite passes.